include/verilated_profiler.h

```diff
@@ -23,58 +23,202 @@
-    size_t counterNumber() const { return m_counterNumber; }
-    std::string name() const { return m_name; }
+    void evalBegin() { m_type = Type::EVAL_BEGIN; }
+    void evalEnd() { m_type = Type::EVAL_END; }
+    void evalLoopBegin() { m_type = Type::EVAL_LOOP_BEGIN; }
+    void evalLoopEnd() { m_type = Type::EVAL_LOOP_END; }
+    void mtaskBegin(vluint32_t id, vluint32_t predictStart) {
+        m_payload.mtaskBegin.id = id;
+        m_payload.mtaskBegin.predictStart = predictStart;
+        m_payload.mtaskBegin.cpu = getcpu();
+        m_type = Type::MTASK_BEGIN;
+    }
+    void mtaskEnd(vluint32_t id, vluint32_t predictCost) {
+        m_payload.mtaskEnd.id = id;
+        m_payload.mtaskEnd.predictCost = predictCost;
+        m_type = Type::MTASK_END;
+    }
 };
 
-// Create some number of bucketed profilers
-template <std::size_t T_Entries> class VerilatedProfiler final {
-    // Counters are stored packed, all together, versus in VerilatedProfilerRec to
-    // reduce cache effects
-    std::array<vluint64_t, T_Entries> m_counters{};  // Time spent on this record
-    std::deque<VerilatedProfilerRec> m_records;  // Record information
+static_assert(std::is_trivially_destructible<VlExecutionRecord>::value,
+              "VlExecutionRecord should be trivially destructible for fast buffer clearing");
+
+//=============================================================================
+// VlExecutionProfiler is for collecting profiling data about model execution
+
+class VlExecutionProfiler final {
+    // TYPES
+
+    // Execution traces are recorded into thread local vectors. We can append records of profiling
+    // events to this vector with very low overhead, and then dump them out later. This prevents
+    // the overhead of printf/malloc/IO from corrupting the profiling data. It's super cheap to
+    // append a VlProfileRec struct on the end of a pre-allocated vector; this is the only cost we
+    // pay in real-time during a profiling cycle. Internal note: Globals may multi-construct, see
+    // verilated.cpp top.
+    using ExecutionTrace = std::vector<VlExecutionRecord>;
+
+    // STATE
+    static VL_THREAD_LOCAL ExecutionTrace t_trace;  // thread-local trace buffers
+    VerilatedMutex m_mutex;
+    // Map from thread id to &t_trace of given thread
+    std::map<uint32_t, ExecutionTrace*> m_traceps VL_GUARDED_BY(m_mutex);
+
+    bool m_enabled = false;  // Is profiling currently enabled
+
+    vluint64_t m_tickBegin = 0;  // Sample time (rdtsc() on x86) at beginning of collection
+    vluint64_t m_lastStartReq = 0;  // Last requested profiling start (in simulation time)
+    vluint32_t m_windowCount = 0;  // Track our position in the cache warmup and profile window
 
 public:
+    // CONSTRUCTOR
+    VlExecutionProfiler();
+
     // METHODS
-    VerilatedProfiler() = default;
-    ~VerilatedProfiler() = default;
+
+    // Is profiling enabled
+    inline bool enabled() const { return m_enabled; }
+
```

Comment:
```suggestion

```
Nit; remove blanks between lines in next few declarations so can more easily see them all in a screenfull.